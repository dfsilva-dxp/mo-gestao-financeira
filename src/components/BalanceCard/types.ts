export type BalanceCardProps = {
  title: string;
  balance: string;
  percentage: string;
  progressBar: number;
  color?: "primary" | "secondary";
};
