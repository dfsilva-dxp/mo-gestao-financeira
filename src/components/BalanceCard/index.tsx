import { TiChartLineOutline } from "react-icons/ti";

import { ProgressBar } from "components";

import {
  Balance,
  CardBody,
  Header,
  Icon,
  Percentage,
  Title,
  Wrapper
} from "./styles";
import { BalanceCardProps } from "./types";

export default function BalanceCard({
  title,
  balance,
  percentage,
  progressBar,
  color = "primary"
}: BalanceCardProps) {
  return (
    <Wrapper>
      <Header>
        <Title>{title}</Title>

        <Icon>
          <TiChartLineOutline />
        </Icon>
      </Header>

      <CardBody>
        <Balance>{balance}</Balance>

        {!!percentage && <Percentage>{percentage}</Percentage>}
      </CardBody>

      <ProgressBar progressBar={progressBar} color={color} />
    </Wrapper>
  );
}
