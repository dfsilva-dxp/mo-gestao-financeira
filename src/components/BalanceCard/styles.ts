import styled, { css, DefaultTheme } from "styled-components";
import { BalanceCardProps } from "./types";

export const Wrapper = styled.div`
  ${({ theme }) => css`
    background: ${theme.colors.gradient};
    padding: ${theme.spacings.small} ${theme.spacings.large};
    border-radius: ${theme.border.radius.default};
  `}
`;

export const Header = styled.header`
  ${({ theme }) => css`
    margin-bottom: ${theme.spacings.xxsmall};
    display: flex;
    align-items: center;
    justify-content: space-between;
  `}
`;

export const Title = styled.h4`
  ${({ theme }) => css`
    color: ${theme.colors.gray};
  `}
`;

export const Icon = styled.div`
  ${({ theme }) => css`
    color: ${theme.colors.gray};
    font-size: ${theme.font.sizes.large};
  `}
`;

export const CardBody = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const Balance = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.large};
  `}
`;

export const Percentage = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.green};
  `}
`;

export const ProgressBar = styled.div`
  ${({ theme }) => css`
    margin-top: ${theme.spacings.xxsmall};
    width: 100%;
  `}
`;

const progressModifiers = {
  primary: (theme: DefaultTheme) => css`
    background: ${theme.colors.primary};
  `,
  secondary: (theme: DefaultTheme) => css`
    background: ${theme.colors.secondary};
  `
};

type ProgressProps = Pick<BalanceCardProps, "progressBar" | "color">;

export const Progress = styled.div<ProgressProps>`
  ${({ theme, progressBar, color }) => css`
    background: ${theme.colors.black};
    width: 100%;
    height: 1rem;
    border-radius: ${theme.border.radius.default};
    position: relative;

    &::after {
      content: "";
      background: ${theme.colors.secondary};
      width: ${`${progressBar}%`};
      height: 1rem;
      border-radius: ${theme.border.radius.default};
      position: absolute;

      ${!!color && progressModifiers[color](theme)}
    }
  `}
`;
