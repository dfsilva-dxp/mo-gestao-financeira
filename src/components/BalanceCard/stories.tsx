import { Story, Meta } from "@storybook/react/types-6-0";
import { ComponentProps } from "react";

import BalanceCard from ".";

export default {
  title: "Components/BalanceCard",
  component: BalanceCard,
  argTypes: {
    color: {
      options: ["primary", "secondary"],
      control: {
        type: "inline-radio"
      }
    }
  },
  args: {
    title: "Saldo",
    balance: "R$8916,07",
    percentage: "+5%",
    progressBar: 80,
    color: "primary"
  },
  parameters: {
    layout: "centered"
  },
  decorators: [
    (Story) => (
      <div style={{ margin: "0 auto", width: "47.6rem" }}>
        <Story />
      </div>
    )
  ]
} as Meta;

export const Basic: Story<ComponentProps<typeof BalanceCard>> = ({
  ...props
}) => <BalanceCard {...props} />;
