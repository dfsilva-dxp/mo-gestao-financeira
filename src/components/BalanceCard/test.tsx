import { screen } from "@testing-library/react";
import { renderWithTheme } from "utils/tests/helpers";

import BalanceCard from ".";

const props = {
  title: "Saldo",
  balance: "R$8916,07",
  percentage: "+5%",
  progressBar: 80
};

describe("<BalanceCard />", () => {
  it("should render the currectly", () => {
    const { container } = renderWithTheme(<BalanceCard {...props} />);

    expect(screen.getByRole("heading", { name: /saldo/i })).toBeInTheDocument();

    expect(screen.getByRole("heading", { name: /saldo/i })).toHaveStyle({
      color: "#797D9A"
    });

    expect(screen.getByText("R$8916,07")).toBeInTheDocument();
    expect(screen.getByText("+5%")).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render primary color by default", () => {
    renderWithTheme(<BalanceCard {...props} />);

    expect(
      screen.getByRole("progressbar", { name: /progress bar/i })
    ).toHaveStyleRule("background", "#F9769D", {
      modifier: "::after"
    });
  });

  it("should render #96A7FF when secondary color is passed", () => {
    renderWithTheme(<BalanceCard {...props} color="secondary" />);

    expect(
      screen.getByRole("progressbar", { name: /progress bar/i })
    ).toHaveStyleRule("background", "#96A7FF", {
      modifier: "::after"
    });
  });
});
