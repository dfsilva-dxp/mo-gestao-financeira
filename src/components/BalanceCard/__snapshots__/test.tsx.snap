// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BalanceCard /> should render the currectly 1`] = `
.c0 {
  background: linear-gradient(145deg,rgba(33,34,45,1) 0%,rgba(58,44,58,1) 100%);
  padding: 2.0rem 4.0rem;
  border-radius: 1.6rem;
}

.c1 {
  margin-bottom: 0.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  color: #797D9A;
}

.c3 {
  color: #797D9A;
  font-size: 2.2rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c5 {
  font-size: 2.2rem;
}

.c6 {
  font-size: 1.4rem;
  color: #3CD3C1;
}

.c7 {
  margin-top: 0.8rem;
  width: 100%;
}

.c8 {
  background: #030517;
  width: 100%;
  height: 1rem;
  border-radius: 1.6rem;
  position: relative;
}

.c8::after {
  content: "";
  background: #96A7FF;
  width: 80%;
  height: 1rem;
  border-radius: 1.6rem;
  position: absolute;
  background: #F9769D;
}

<div
  class="c0"
>
  <header
    class="c1"
  >
    <h4
      class="c2"
    >
      Saldo
    </h4>
    <div
      class="c3"
    >
      <svg
        baseProfile="tiny"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        version="1.2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5.999 17c-.677 0-1.342-.234-1.873-.658-.626-.501-1.019-1.215-1.107-2.011-.089-.796.138-1.58.639-2.206l4-5c.978-1.225 2.883-1.471 4.143-.523l1.674 1.254 2.184-2.729c.571-.716 1.425-1.127 2.342-1.127.679 0 1.343.232 1.873.657.626.501 1.021 1.216 1.108 2.013s-.14 1.58-.641 2.204l-4 5c-.977 1.226-2.882 1.471-4.143.526l-1.674-1.256-2.184 2.729c-.57.716-1.424 1.127-2.341 1.127zm4.001-9c-.306 0-.591.137-.781.374l-4 5.001c-.167.208-.243.471-.213.734.03.266.161.504.369.67.228.183.465.221.624.221.306 0 .591-.137.782-.376l3.395-4.244 3.224 2.42c.42.316 1.056.231 1.381-.176l4-5.001c.167-.208.242-.469.213-.734-.03-.266-.161-.504-.369-.67-.227-.182-.464-.22-.624-.22-.306 0-.591.137-.782.376l-3.395 4.242-3.224-2.417c-.175-.132-.382-.2-.6-.2zM19 21h-14c-.552 0-1-.447-1-1s.448-1 1-1h14c.552 0 1 .447 1 1s-.448 1-1 1z"
        />
      </svg>
    </div>
  </header>
  <div
    class="c4"
  >
    <p
      class="c5"
    >
      R$8916,07
    </p>
    <p
      class="c6"
    >
      +5%
    </p>
  </div>
  <div
    class="c7"
  >
    <div
      aria-label="Progress Bar"
      class="c8"
      color="primary"
      role="progressbar"
    />
  </div>
</div>
`;
